File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const input = require('fs').readFileSync(0).toString().trim().split('\n');

const [N, M] = input.shift().split(' ').map(Number);
const maze = input.map((v) => v.split('').map(Number));
const [goalY, goalX] = [N - 1, M - 1]; 
const ds = [[-1, 0], [1, 0], [0, 1], [0, -1]]; 

const queue = [[0, 0, 1]]; 

while (queue.length) {
  const [curY, curX, move] = queue.shift();

  if (goalY === curY && goalX === curX) console.log(move);

  for (let i = 0; i < 4; i++) {
    const ny = curY + ds[i][1];
    const nx = curX + ds[i][0];

    if (ny >= 0 && ny < N && nx >= 0 && nx < M && maze[ny][nx]) {
        maze[ny][nx] = 0; 
        queue.push([ny, nx, move + 1]); 
    }
  }
};

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
def bfs(x, y):
    queue = [(x, y)]

    while queue:
        x, y = queue.pop(0)
        for i in range(4):
            nx = x + dir[i][0]
            ny = y + dir[i][1]
            if 0 <= nx < N and 0 <= ny < M and maze[nx][ny] == 1:
                maze[nx][ny] = maze[x][y] + 1
                queue.append((nx, ny))

N, M = map(int, input().split())
maze = [list(map(int, input().strip())) for _ in range(N)]
dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]

bfs(0, 0)

print(maze[N - 1][M - 1])